File: src/ru/edu/vtisov/array/BestTimeToBuyAndSellStockWithTransactionFee.java
package ru.edu.vtisov.array;

// https://leetcode.com/problems/best-time-to-buy-and-sell-stock-with-transaction-fee/

public class BestTimeToBuyAndSellStockWithTransactionFee {

    public int maxProfit(int[] prices, int fee) {
        int cash = 0, hold = -prices[0];
        for (int i = 1; i < prices.length; i++) {
            cash = Math.max(cash, hold + prices[i] - fee);
            hold = Math.max(hold, cash - prices[i]);
        }
        return cash;
    }
}

File: src/ru/edu/vtisov/dp/best-time-to-buy-and-sell-stock-with-transaction-fee.go
package dp

func maxProfit(prices []int, fee int) int {
	hold := make([]int, len(prices))
	free := make([]int, len(prices))

	free[0] = 0
	hold[0] = -prices[0]

	for i := 1; i < len(prices); i++ {
		free[i] = max(free[i-1], hold[i-1]+prices[i]-fee)
		hold[i] = max(hold[i-1], free[i-1]-prices[i])
	}
	return free[len(prices)-1]
}
